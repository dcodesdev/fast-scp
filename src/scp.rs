use anyhow::Ok;
use ssh2::Session;
use std::{
    fs::{self, File},
    io::{Read, Write},
    net::TcpStream,
    path::{Path, PathBuf},
};

pub struct Receiver {
    /// Destination directory
    dest: PathBuf,
    session: Session,
}

impl Receiver {
    pub fn new(ssh_opts: SshOpts) -> anyhow::Result<Self> {
        Ok(Self {
            dest: PathBuf::from("."),
            session: create_session(&ssh_opts)?,
        })
    }

    pub fn dir(mut self, dir: PathBuf) -> Self {
        self.dest = dir;
        self
    }

    pub fn receive(&self, path: &PathBuf) -> anyhow::Result<()> {
        if self.is_dir(path)? {
            self.handle_dir(path)
        } else {
            copy_file_from_vps(
                &self.session,
                path,
                &self.dest.join(path.file_name().unwrap()),
            )
        }
    }

    fn handle_dir(&self, dir: &PathBuf) -> anyhow::Result<()> {
        let dirs = list_dir(&self.session, &dir);

        for item in dirs.iter() {
            println!("File: {:?}", item);
        }

        unimplemented!("Handle dir")
    }

    fn is_dir(&self, path: &PathBuf) -> anyhow::Result<bool> {
        let sftp = self.session.sftp().unwrap();
        let metadata = sftp.stat(path)?;

        Ok(metadata.is_dir())
    }
}

pub struct SshOpts {
    pub host: String,
    pub username: String,
    pub private_key: PathBuf,
}

fn copy_file_from_vps(
    session: &Session,
    remote_file_path: &PathBuf,
    local_file_path: &PathBuf,
) -> anyhow::Result<()> {
    // Create a SCP channel for receiving the file
    let (mut remote_file, stat) = session.scp_recv(&remote_file_path)?;
    let mut contents = Vec::with_capacity(stat.size() as usize);
    remote_file.read_to_end(&mut contents)?;

    // make the dir if not exists
    fs::create_dir_all(local_file_path.parent().unwrap())?;

    // Create local file and write to it
    let mut local_file = File::create(local_file_path)?;
    local_file.write_all(&contents)?;

    Ok(())
}

fn list_dir(session: &Session, dir: &Path) -> anyhow::Result<Vec<PathBuf>> {
    let sftp = session.sftp()?;
    let dir = sftp.readdir(dir)?;
    let dirs = dir.into_iter().map(|entry| entry.0).collect();
    Ok(dirs)
}

pub fn create_session(ssh_opts: &SshOpts) -> anyhow::Result<Session> {
    // Connect to the host
    let tcp = TcpStream::connect(&ssh_opts.host)?;
    let mut session = Session::new()?;
    session.set_tcp_stream(tcp);
    session.handshake()?;

    // Authenticate using a private key
    session.userauth_pubkey_file(&ssh_opts.username, None, &ssh_opts.private_key, None)?;

    Ok(session)
}
