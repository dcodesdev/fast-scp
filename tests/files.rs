use std::fs;

fn open_file() -> String {
    fs::read_to_string("tests/snapshots/ls-R.log").expect("Unable to read file")
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta;
    use scp_rs::scp::*;

    #[test]
    fn test_find_files() {
        let file = open_file();
        let files_only = find_files(&file);
        println!("{:?}", files_only);
        insta::assert_debug_snapshot!(files_only);
    }
}
