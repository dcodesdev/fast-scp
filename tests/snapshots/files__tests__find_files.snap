---
source: tests/files.rs
assertion_line: 18
expression: files_only
---
[
    "project/README.md",
    "project/package.json",
    "project/src/app/App.tsx",
    "project/src/app/index.ts",
    "project/src/components/Header.tsx",
    "project/src/components/Footer.tsx",
    "project/src/components/Sidebar.tsx",
    "project/src/components/index.ts",
    "project/src/hooks/useAuth.ts",
    "project/src/hooks/useFetch.ts",
    "project/src/hooks/index.ts",
    "project/src/utils/helpers.ts",
    "project/src/utils/constants.ts",
    "project/src/utils/index.ts",
    "project/tests/App.test.tsx",
    "project/tests/Header.test.tsx",
    "project/tests/Footer.test.tsx",
    "project/tests/Sidebar.test.tsx",
]
